lib/Core/Memory.h

```diff
@@ -120,6 +120,15 @@ class MemoryObject {
   ref<Expr> getOffsetExpr(const ref<Expr> &pointer) const {
     return SubExpr::create(pointer, getBaseExpr());
   }
+
+  /// Return the offset of the object the pointer is pointing to.
+  /// \param pointer
+  /// \return offset with respect to this object
+  uint64_t getOffset(uint64_t pointer) const {
```

Comment:
I would overload `getOffsetExpr` and template the snippet in Executor.

Reply:
Just out of curiosity, how would you template that snippet?
The algorithmic structure is different.
There is currently nothing in place in KLEE which allows to fold arbitrary `Expr` or `ref<Expr>` into numeric values at compile time.

By the way, the function you reference has an `Expr` in its name, so I would expect an `Expr` as result.

Reply:
Something like this, but might be too optimistic. 

```C++
template <typename T> foo(T address, ...) {
    auto offset = mo->getOffsetExpr(address);
    auto check = mo->getBoundsCheckOffset(offset, bytes);

//    check = optimizer.optimizeExpr(check, true);

    bool inBounds = false;
    if(isa<Expr>(address)) {
    solver->setTimeout(coreSolverTimeout);
    bool success = solver->mustBeTrue(state, check, inBounds);
    solver->setTimeout(time::Span());
    if (!success) {
      state.pc = state.prevPC;
      terminateStateEarly(state, "Query timed out (bounds check).");
      return;
    } } else {
    inBounds = check
   }


    if (inBounds) {
      const ObjectState *os = op.second;
      if (isWrite) {
        if (os->readOnly) {
          terminateStateOnError(state, "memory error: object read only",
                                ReadOnly);
        } else {
          ObjectState *wos = state.addressSpace.getWriteable(mo, os);
          wos->write(offset, value);
        }          
      } else {
        ref<Expr> result = os->read(offset, type);
        
        if (interpreterOpts.MakeConcreteSymbolic)
          result = replaceReadWithSymbolic(state, result);
        
        bindLocal(target, state, result);
      }

      return;
}
```

I really don't think this code duplication is helping anyone, the "algorithmic difference" is tiny.

You could always rename the `getOffsetExpr` to just `getOffset`.

Reply:
Ok, I see. Unfortunately, this doesn't solve the problem at all.

Remember, the original issue that we try to solve here is that concrete accesses are just too slow. Now imagine, you've got a memory intensive operation like `memcpy`, `memset`, ... . In a tight loop, it reads/write every single byte one by one. For every byte, in your version, you create a `ConstantExpr`, calculate the offset by invoking `SubExpr::create`, which checks for constantness, at the end substract the value and create a result expression. Each expression is allocated on heap. And up to now, I just described your two first lines. There is no solver invocation and bounds-check involved that happens later.

Sure, the code of this PR does on an abstract level the same, but does it with concrete values directly that do not involve creating expressions or invoking the solver in the first place.

Reply:
Surw, but if you instantiate `T` to `uint64_t`, you get your code, and the solver call is skipped. Atleast that's the idea.

Reply:
Not only this, you especially skip creating multiple expressions in the first place where none are needed.

Reply:
yes, so would this templated version with the integer instantiation and the right overloads,

Reply:
Nope. Of course, you can prove me wrong. As a start, just try to come up with a real, proper template for the `getOffsetExpr()`

Reply:
You don't need a template for `getOffsetExpr`, you can just have `uint64_t getOffsetExpr(uint64_t)` and `ref<Expr> getOffsetExpr( ref<Expr>)` overloads.

Reply:
One returns a value the other an expression.

Reply:
yes, isn't that the point of your optimisation?

Reply:
@MartinNowack since the argument list is different the changing the return type is allowed. I agree though that you aren't gaining much since the caller still is going to need to special case. I think it principle what @kren1 describes is possible but I don't know that it is less complex, you still need a bunch of overloads/specialized-template-functions just to have a single code path higher up.